File: core/src/com/ultreon/craft/registry/DelayedRegister.java
package com.ultreon.craft.registry;

import com.ultreon.craft.event.RegistryEvents;
import com.ultreon.craft.resources.Identifier;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

@Deprecated
public class DelayedRegister<T> {
    private final String namespace;
    private final Registry<T> registry;
    private final Map<Identifier, Supplier<? extends T>> deffer = new HashMap<>();

    public DelayedRegister(String namespace, Registry<T> registry) {
        this.namespace = namespace;
        this.registry = registry;
    }

    public <C extends T> RegistrySupplier<C> register(String name, Supplier<C> supplier) {
        Identifier id = new Identifier(namespace, name);
        deffer.put(id, supplier);
        return new RegistrySupplier<>(registry, id);
    }

    public void register() {
        RegistryEvents.get(registry).listen(reg -> {
            for (var entry : deffer.entrySet())
                reg.set(entry.getKey(), entry.getValue().get());
        });
    }
}

File: src/jvmMain/kotlin/com/ultreon/craft/registry/DelayedRegister.kt
package com.ultreon.craft.registry

import com.ultreon.craft.event.*
import com.ultreon.craft.resources.*

class DelayedRegister<T : Any>(
    private val namespace: String,
    private val registry: Registry<T>
) {
    private val deffer: MutableMap<String, () -> T> = mutableMapOf()

    init {
        RegistryEvents[registry.type] += { reg ->
            for ((name, func) in deffer) {
                reg[id(namespace, name)] = func()
            }
        }
    }

    fun <C : T>register(name: String, func: () -> C): RegistrySupplier<T> {
        deffer[name] = func
        return RegistrySupplier(registry, id(namespace, name))
    }
}
